### PR TITLE
[cuBLASLt][Memtracker] Allocate temprorary cuBLASLt workspaces using tensors rather than going to the caching allocator directly 

### PR DESCRIPTION
**Pull request by eqy**
CC @zdevito @janeyx99 

This isn't ideal but cuBLASLt workspaces are not currently cached, so this additional untracked allocation will cause `test_cuda_tracker_equivalence` to fail with a large enough workspace size e.g., `CUBLAS_LT_WORKSPACE_SIZE=32768`. One solution is to just use byte-tensors for the workspace instead of going directly to the caching allocator.

cc @ptrblck @msaroufim @csarofeen @xwang233

### CODE REVIEW

**Comment by eqy** [comment]
CC @nWEIdia @tinglvv as we discussed this

@Aidyn-A as 32MiB default workspace size for H100 is relevant here

**Review by Aidyn-A — APPROVED** [review]
Looks good to me. QQ: does `at::empty` has the `nullptr` check? If yes, this:
```C++
TORCH_CHECK(workspace.data_ptr() != nullptr, "OOM trying to allocate workspace for cublaslt");
```
would be redundant.

**Comment by janeyx99** [comment]
Does this have performance implications? Like is at::empty overhead going to make this slower than it was before to call cuBlas APIs?

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 1]
```diff
@@ -1468,9 +1466,8 @@ void scaled_gemm(
     computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_BIAS_DATA_TYPE, ScalarTypeToCudaDataType(bias_dtype));
   }
   size_t workspaceSize = _getWorkspaceSize();
-  auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
-  auto workspace = allocator.allocate(workspaceSize);
-  TORCH_CHECK(workspace.get() != nullptr, "OOM trying to allocate workspace for cublaslt");
+  auto workspace = at::empty(workspaceSize, at::TensorOptions().dtype(at::kByte).device(at::kCUDA));
+  TORCH_CHECK(workspace.data_ptr() != nullptr, "OOM trying to allocate workspace for cublaslt");
```
Comment (Skylion007) [inline aten/src/ATen/cuda/CUDABlas.cpp:1470]:
```suggestion
  TORCH_CHECK(workspace.const_data_ptr() != nullptr, "OOM trying to allocate workspace for cublaslt");
```
?

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 2]
```diff
@@ -1554,7 +1551,7 @@ void scaled_gemm(
       result_ptr,
       Ddesc.descriptor(),
       &heuristicResult.algo,
-      workspace.mutable_get(),
+      workspace.data_ptr(),
```
Comment (Skylion007) [inline aten/src/ATen/cuda/CUDABlas.cpp:1554]:
```suggestion
      workspace.mutable_data_ptr(),
```
?

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 3]
```diff
@@ -1664,7 +1661,7 @@ void int8_gemm(
       nullptr, // Heuristics don't seem to work for int8
 #endif
 #ifdef USE_ROCM
-      workspace.mutable_get(),
+      workspace.data_ptr(),
```
Comment (Skylion007) [inline aten/src/ATen/cuda/CUDABlas.cpp:1664]:
```suggestion
      workspace.mutable_data_ptr(),
```

Comment (Skylion007) [reply aten/src/ATen/cuda/CUDABlas.cpp:1664]: etc... I think these are mostly for robust COW behavior of tensors, but no reason not to be explicit here.

**Comment by eqy** [comment]
> Does this have performance implications? Like is at::empty overhead going to make this slower than it was before to call cuBlas APIs?

Did some microbenchmarking finally, seems like the overhead is in the range of 0.5us
```
#include <torch/torch.h>
#include <torch/cuda.h>
#include <c10/cuda/CUDACachingAllocator.h>

#include <iostream>
#include <chrono>

#define ITER 1000000

int main() {
  auto caching_allocator = c10::cuda::CUDACachingAllocator::get();
  std::cout << "start empty" << std::endl;
  auto t = at::empty({1024}, at::device(at::kCUDA));
  torch::cuda::synchronize();
  std::cout << "empty warmup finished" << std::endl;
  auto t0 = std::chrono::high_resolution_clock::now();
  for (int i = 0; i < ITER; i++) {
    t = at::empty({1024}, at::device(at::kCUDA));
  }
  torch::cuda::synchronize();
  auto t1 = std::chrono::high_resolution_clock::now();
  std::cout << "start allocate" << std::endl;
  auto ptr = caching_allocator->allocate(1024*4);
  torch::cuda::synchronize();
  std::cout << "allocate warmup finished" << std::endl;
  auto t2 = std::chrono::high_resolution_clock::now();
  for (int i = 0; i < ITER; i++) {
    ptr = caching_allocator->allocate(1024*4);
  }
  torch::cuda::synchronize();
  auto t3 = std::chrono::high_resolution_clock::now();
  auto empty_time = std::chrono::duration_cast<std::chrono::duration<double>>(t1 - t0).count();
  auto allocate_time = std::chrono::duration_cast<std::chrono::duration<double>>(t3 - t2).count();
  std::cout << "empty time per iter: " << empty_time/ITER << std::endl;
  std::cout << "allocate time per iter: " << allocate_time/ITER << std::endl;
}
```

```
start empty
empty warmup finished
start allocate
allocate warmup finished
empty time per iter: 7.88432e-07
allocate time per iter: 2.3696e-07
```

**Review by albanD — APPROVED** [review]
SGTM

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**